core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -545,61 +546,117 @@ object Validator {
+        failValidation(
+          s"message",
+          "explanatory"
         )
+      case Some(fields) => validateFields(fields)
+    }
+  }
 
-    def doesNotStartWithUnderscore(field: __InputValue): IO[ValidationError, Unit] =
-      IO.when(field.name.startsWith("__"))(
+  private def onlyInputType(`type`: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsInputType()
+    def isInputType(t: __Type): Either[__Type, Unit] = {
+      import __TypeKind._
+      t.kind match {
+        case LIST | NON_NULL              => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
+        case SCALAR | ENUM | INPUT_OBJECT => Right(())
+        case _                            => Left(t)
+      }
+    }
+
+    IO.whenCase(isInputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject can't start with '__': ${field.name}",
-          """The input field must not have a name which begins with the
-characters {"__"} (two underscores)"""
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
+          """The input field must accept a type where IsInputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      )
+    }
+  }
 
-    def onlyInputFieldType(field: __InputValue): IO[ValidationError, Unit] =
-      IO.whenCase(isInputType(field.`type`())) {
-        case Left(errorType) =>
-          failValidation(
-            s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
-            """The input field must accept a type where IsInputType(inputFieldType) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
-          )
+  private def onlyOutputType(`type`: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsOutputType()
+    def isOutputType(t: __Type): Either[__Type, Unit] = {
+      import __TypeKind._
+      t.kind match {
+        case LIST | NON_NULL                            => t.ofType.fold[Either[__Type, Unit]](Left(t))(isOutputType)
+        case SCALAR | OBJECT | INTERFACE | UNION | ENUM => Right(())
+        case _                                          => Left(t)
       }
+    }
 
-    t.inputFields match {
-      case None | Some(Nil) =>
+    IO.whenCase(isOutputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject ${t.name.getOrElse("")} does not have fields",
-          "An Input Object type must define one or more input fields"
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an OutputType",
+          """The input field must accept a type where IsOutputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      case Some(fields) => validateFields(fields)
     }
   }
 
+  private def noDuplicateFieldName(fields: List[__Field]) =
+    noDuplicateName[__Field](fields, _.name)
```

Comment:
Is it necessary to set the type explicitly?

Reply:
`noDuplicateName(fields, _.name)' and `noDuplicateName(fields, arg => arg.name)'  results in "missing parameter type"

`noDuplicateName(fields, (arg: __Field) => arg.name)` is compiling fine but longer than explicitly setting type parameter T on the method

I actually don't know why the compiler can't infer the type for T. Is there a better way to do it?

Reply:
Ah OK, that's not a big deal. I thought the compiler would infer it but if it doesn't, no problem with leaving it.